olive/engine/footprint.py

```diff
@@ -0,0 +1,178 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+import logging
+from collections import OrderedDict
+from datetime import datetime
+from pathlib import Path
+
+from olive.common.config_utils import ConfigBase, config_json_dumps, config_json_loads
+
+logger = logging.getLogger(__name__)
+
+
+class FootprintNodeMetric(ConfigBase):
+    value: dict
+    cmp_direction: dict = {}
+    is_goals_met: bool = False
+
+
+class FootprintNode(ConfigBase):
+    # None for no parent which means current model is the input model
+    parent_model_id: str = None
+    model_id: str
+    model_config: dict = None
+    from_pass: str = None
+    pass_run_config: dict = None
+    is_pareto_frontier: bool = False
+    metrics: FootprintNodeMetric = FootprintNodeMetric(value={}, is_goals_met=False)
+
+    date_time: float = datetime.now().timestamp()
+
+    def update(self, **kwargs):
+        for k, v in kwargs.items():
+            setattr(self, k, v)
+
+
+class Footprint:
+    """
+    The engine footprint is a class that contains the footprint of the engine runtime.
+    It is used to collect the runtime state of the Olive engine and to organize the state
+    in a way that is easy to visualize and understand for customers.
+    """
+
+    def __init__(
+        self,
+        nodes: OrderedDict = {},
```

Comment:
>{} [](http://example.com/codeflow?start=29&length=2)

please don't use the empty dict or empty list as the default argument. Instead, using None.

[https://stackoverflow.com/questions/366422/how-can-i-avoid-issues-caused-by-pythons-early-bound-default-parameters-e-g-m](https://stackoverflow.com/questions/366422/how-can-i-avoid-issues-caused-by-pythons-early-bound-default-parameters-e-g-m)